lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,263 @@
+	switch connectionOptions.mode {
+	case Standalone:
+		redisConn = redis.NewClient(&redis.Options{
+			Addr:      connectionAddr,
+			TLSConfig: tlsConfig,
+		})
+	case Cluster:
+		redisConn = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs:     []string{connectionAddr},
+			TLSConfig: tlsConfig,
+		})
+	default:
+		// We've checked that while validating the config, but checking again can help with regression.
+		return trace.BadParameter("incorrect connection mode %s", connectionOptions.mode)
+	}
+
+	defer func() {
+		if err := redisConn.Close(); err != nil {
+			e.Log.Errorf("failed to close Redis connection: %v", err)
+		}
+	}()
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	if err := e.process(ctx, redisConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// process is the main processing function for Redis. It reads commands passed from client and passes them to
+// a Redis instance. It's also responsible for audit.
+func (e *Engine) process(ctx context.Context, redisClient redis.UniversalClient) error {
+	for {
+		// Read commands from client.
+		cmd, err := e.readClientCmd(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// send valid commands to Redis instance/cluster.
+		err = e.processCmd(ctx, redisClient, cmd)
+
+		// extract server response from
+		vals, err := serverResponseToValue(cmd, err)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Send response back to the client.
+		if err := e.sendToClient(vals); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// readClientCmd reads commands from connected Redis client.
+func (e *Engine) readClientCmd(ctx context.Context) (*redis.Cmd, error) {
+	cmd := &redis.Cmd{}
+	if err := cmd.ReadReply(e.clientReader); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	val, ok := cmd.Val().([]interface{})
+	if !ok {
+		return nil, trace.BadParameter("failed to cast Redis value to a slice, got %T", cmd.Val())
+	}
+
+	return redis.NewCmd(ctx, val...), nil
+}
+
+// serverResponseToValue takes server response and an error and return a message that should be propagated back
+// to connected client as the first argument and errors that should terminate the connection as the second one.
+// Some examples:
+// * Server errors are returned as the first argument. They should be propagated back to the connected client
+//   instead of handled by Teleport.
+// * Go's context.DeadlineExceeded is returned as the first argument as "connection timeout", as returning
+//   context deadline exceeded to a user is confusing.
+// * Connection errors (EOF, closed network) are returned as an errors, as in this case we want to terminate the client connection.
+func serverResponseToValue(cmd *redis.Cmd, err error) (interface{}, error) {
+	var vals interface{}
+	if _, ok := err.(redis.Error); ok {
+		vals = err
+	} else if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
nit: I know you've refactored this already, so NBD, but did you consider using a switch statement here instead of if/else? Switch in this case would be a bit easier to read IMO (fewer curly brackets). :)

Reply:
I simplified the logic a bit. I think now it's cleaner.